### ISSUE TITLE
Return type of function call incorrectly typed as `unknown`

### ISSUE THREAD

**Issue by Diggsey** [microsoft/TypeScript#52489]
# Bug Report

### 🔎 Search Terms

unknown function call ReturnType

### 🕗 Version & Regression Information

All versions <= 4.9.4

- This is the behavior in every version I tried, and I reviewed the FAQ for related entries

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.4#code/FAMwrgdgxgLglgewgAjhAbgg1gUwDwBiyOAHjDhACYDOyAFAHRMCGATgObUBcyEO6OVgG0AugEpkAXgB8ySFggIA7hGl0QPAgBpkTBm048ACm2YBbHOVbVC0sTwBKlsKwgAVAJ4AHfAVkBvYGRg5FZnV2QQRhYOajFgAF9gUEhYRBRyahg6Eh4IMDMAI0F7ZCzWNHZkQJDQ8JQSBhgEAGUYCoh2Onik4DRMXDpMmB0ARnigA)

### 💻 Code

```ts
function invoke<F extends (...args: never[]) => unknown>(f: F, ...args: Parameters<F>): ReturnType<F> {
    return f(...args)
}

function test(x: number): string {
    return x.toString()
}

invoke(test, 1)
```

### 🙁 Actual behavior

> Type 'unknown' is not assignable to type 'ReturnType<F>'.

### 🙂 Expected behavior

No errors.

The result of calling `f(..args)` should always be compatible with `ReturnType<F>`.

**Comment by MartinJohns** [comment]
Duplicate of #33912.

**Comment by ahejlsberg** [comment]
This is working as intended. The type of `f` is not specific enough and `f` is callable only once its type is degraded to its constraint (at which point the return type is simply `unknown`). It works if you give `f` a more specific type:

```ts
function invoke<A extends unknown[], R>(f: (...args: A) => R, ...args: A): R {
    return f(...args);
}

function test(x: number): string {
    return x.toString();
}

invoke(test, 1);  // Returns type string
```

**Comment by Diggsey** [comment]
> `f` is callable only once its type is degraded to its constraint

Why? It doesn't make sense to me that `ReturnType<F>` does not require `F` to be degrated to its constraint, but actually calling `f()` does.

I would expect `ReturnType<F>` to be exactly equivalent to `typeof f(...args)` the same way I would expect the other utility types (like Awaited<T>) to be equivalent to their corresponding expressions.

**Comment by fatcerberus** [comment]
Conditional types (like `ReturnType`) are deferred in a generic context because it's not known by the compiler what the final input type will be.  It's effectively a black box at that point; the compiler doesn't see it in higher-order terms like "return type of `F`" (there's no first-class type op for that) but as just another unresolved conditional type.  On the other hand, calling `f` *does* require `F` to be resolved to a concrete type in order for the compiler to extract its return type (so it can check the rest of the function), and the only type it can use there is the constraint.

**Comment by Diggsey** [comment]
> Conditional types (like ReturnType) are deferred in a generic context because it's not known by the compiler what the final input type will be. It's effectively a black box at that point; the compiler doesn't see it in higher-order terms like "return type of F" (there's no first-class type op for that) but as just another unresolved conditional type.

This makes sense to me although the distinction seems a little academic (an unresolved conditional type *is* a form of higher kinded type).

> On the other hand, calling f does require F to be resolved to a concrete type in order for the compiler to extract its return type (so it can check the rest of the function)

This is where I am struggling to follow.

The compiler is clearly capable of checking the types of variables which do not have concrete types. eg.

```ts
type ReturnType1<T extends (...args: never[]) => unknown> = T extends (...args: never[]) => infer R ? R : never
type ReturnType2<T extends (...args: never[]) => unknown> = T extends (...args: never[]) => infer R ? R : never

function adapt<F extends (...args: never[]) => unknown>(arg: unknown): ReturnType1<F> {
    return arg as ReturnType1<F>
}

function invoke<F extends (...args: never[]) => unknown>(f: F, ...args: Parameters<F>): ReturnType2<F> {
    // The compiler is able to determine here that the two unresolved types ReturnType1<F> and ReturnType2<F>
    // are equivalent, despite neither type being resolvable to a concrete type.
    return adapt(f(...args))
}

function test(x: number): string {
    return x.toString()
}

invoke(test, 1)
```

I don't see why the compiler needs to extract a concrete type for the return value of `f()` when it is evidently capable of type-checking conditional types without resolving them.

**Comment by fatcerberus** [comment]
> This makes sense to me although the distinction seems a little academic (an unresolved conditional type is a form of higher kinded type).

This is true, but the distinction *does* matter in practice since the compiler isn’t capable of reasoning at that level of abstraction, thus the type is treated as opaque.  This kind of analysis is actually harder than it sounds because [conditional types are Turing complete](https://github.com/microsoft/TypeScript/issues/14833), and the type system being structural muddies the waters even further (users would reasonably expect any kind of higher-kinded type checking to respect the structure of the output types).

> I don't see why the compiler needs to extract a concrete type for the return value of `f()` when it is evidently capable of type-checking conditional types without resolving them.

Because there’s no first-class type operator a la `returnof F` by which the compiler could keep the return value of the function generic.  I suppose TS could say the return type is `ReturnType<F>` but that feels a bit weird since it’s defined in userland.  (It would also be a massive breaking change since there’s likely code in the wild depending on the reduction-to-constraint behavior.)

**Comment by fatcerberus** [comment]
fwiw the type checking, such as it is, between unresolved conditionals is pretty rudimentary: it’s essentially a dumb pseudo-nominal check that asks “are the `extends` type and true and false types identical?  If yes, the types are compatible, otherwise no” (note: “identical” is a much stronger constraint than assignability).

**Comment by Diggsey** [comment]
> This kind of analysis is actually harder than it sounds

No doubt, I don't really expect it to handle really complicated conditionals, but it already seems capable enough to cope with the examples here.

> I suppose TS could say the return type is ReturnType<F> but that feels a bit weird since it’s defined in userland.

The compiler can just have its own definition of `ReturnType<F>` which happens to be structurally compatible with the normal "userland" one. I don't think that's too weird.

> It would also be a massive breaking change since there’s likely code in the wild depending on the reduction-to-constraint behavior.

Would it? For this particular case, it seems like it should be backwards compatible. In non-generic situations, there is no difference, and in generic cases, the "reduced-to-constraint return type" must always be a super-type of `ReturnType<F>` otherwise the constraint would be violated. I don't see how this could lead to breakage, since everything is nicely covariant?

> wiw the type checking, such as it is, between unresolved conditionals is pretty rudimentary: it’s essentially a dumb pseudo-nominal check that asks “are the extends type and true and false types identical?

It seems to be slightly cleverer around that wrt `any`, but yeah. It still seems easily "good enough" for this use-case. One could also imagine improving this in future (eg. you could check if the "true" type is a subtype of the "false" type or vice versa, and relax the check in that case from "identical" to a covariance/contravariance check).

**Comment by fatcerberus** [comment]
The breakage I’m envisioning is that a deferred `ReturnType<F>` would not be assignable to any concrete type, so code that assumes it can treat a return value as e.g. ‘number` because the function is constrained to a type that returns `number` would no longer work.

**Comment by Diggsey** [comment]
If `F` is constrained to `(...args: never[]) => number` (which is the minimal possible constraint for such code to be valid) then `ReturnType<F>` is known to be a subclass of `number`. This logic already exists in the compiler, since this already typechecks:

```ts
type ReturnType1<T extends (...args: never[]) => unknown> = T extends (...args: never[]) => infer R ? R : never

function giveMeAn<X>(): X {
    throw "Not implemented"
}

function test<F extends (...args: never[]) => number>(f: F) {
    const x: number = giveMeAn<ReturnType1<F>>()
}
```

**Comment by fatcerberus** [comment]
Ah, thanks, I wasn’t aware of that behavior.  That’s interesting.
